Federated tracing

cappuc opened the pull request:
Resolves #2053 
Resolves #2068 
 
- [x] Added or updated tests
- [ ] Documented user facing changes
- [ ] Updated CHANGELOG.md

**Changes**

This PR adds support for [Apollo Federated tracing](https://www.apollographql.com/docs/federation/metrics/).

I used the [apollo server](https://github.com/apollographql/apollo-server/blob/main/packages/server/src/plugin/inlineTrace/index.ts) implementation as reference.

**Error reporting**

Federated tracing spec supports errors reporting but we need to decide how to implement this feature in order to prevent exposing sensitive data (without user control), **right now there is no setting for this**.

Apollo server has this options:
- mask error (default)
- include the full error
- transform the error with a callback

Here the [docs](https://www.apollographql.com/docs/apollo-server/api/plugin/inline-trace#options) and [implementation](https://github.com/apollographql/apollo-server/blob/c1105b3b566b293c849922c7f600cc6b4f5f8ecb/packages/server/src/plugin/traceTreeBuilder.ts#L43C1-L53C51)

**Breaking changes**

None. the old tracing implementation is enabled by default.
This new implementation must be manually enabled through config.

**Some notes**

The new tracing spec uses `protobuf` and the updated of the proto file is manual:
- copy the `reports.proto` file from the [apollo-server repo](https://github.com/apollographql/apollo-server/blob/main/packages/usage-reporting-protobuf/src/reports.proto)
- remove the js specific directives (for example: `[(js_use_toArray)=true]` and `[(js_preEncoded)=true]`
- add the php options:
   ```proto
   option php_namespace = "Nuwave\\Lighthouse\\Tracing\\FederatedTracing\\Proto";
   option php_metadata_namespace = "Nuwave\\Lighthouse\\Tracing\\FederatedTracing\\Proto\\Metadata";
   ```
Then run `make:proto` and `make:fix` to generate (and fix sytle) the protobuf classes

spawnia commented on composer.json:
```diff
@@ -38,6 +38,7 @@
         "illuminate/routing": "^9 || ^10",
         "illuminate/support": "^9 || ^10",
         "illuminate/validation": "^9 || ^10",
+        "google/protobuf": "^3.21",
```
Let's add that to the dev-dependencies and to `suggest` as well, adding new deps is a breaking change.

spawnia commented on src/Events/BuildExtensionsResponse.php:
```diff
@@ -2,10 +2,20 @@
+class BuildExtensionsResponse
+{
+    public function __construct(
+        /**
+         * The result of resolving a single operation.
+         */
```
```suggestion
        /** The result of resolving a single operation. */
```

spawnia commented on src/Tracing/ApolloTracing/ApolloTracing.php:
```diff
@@ -0,0 +1,78 @@
+use Nuwave\Lighthouse\Tracing\Tracing;
+use Nuwave\Lighthouse\Tracing\TracingUtilities;
+
+/**
+ * See https://github.com/apollographql/apollo-tracing#response-format.
+ */
```
```suggestion
/** See https://github.com/apollographql/apollo-tracing#response-format. */
```

spawnia commented on src/Tracing/FederatedTracing/FederatedTracing.php:
```diff
@@ -0,0 +1,194 @@
+use Nuwave\Lighthouse\Tracing\Tracing;
+use Nuwave\Lighthouse\Tracing\TracingUtilities;
+
+/**
+ * See https://www.apollographql.com/docs/federation/metrics/.
+ */
```
```suggestion
/** See https://www.apollographql.com/docs/federation/metrics. */
```

spawnia commented on src/Tracing/FederatedTracing/FederatedTracing.php:
```diff
@@ -0,0 +1,194 @@
+
+    protected Trace $trace;
+
+    protected float|int $requestStartPrecise;
```
Unit?

spawnia commented on src/Tracing/FederatedTracing/FederatedTracing.php:
```diff
@@ -0,0 +1,194 @@
+        assert($this->trace !== null);
+
+        return new ExtensionsResponse(
+            'ftv1',
```
duplicated magic value

cappuc commented on src/Tracing/TracingServiceProvider.php:
```diff
@@ -25,7 +36,8 @@ public function boot(Dispatcher $dispatcher): void
             $manipulateAST->documentAST,
             Parser::constDirective('@tracing'),
         ));
-        $dispatcher->listen(StartExecution::class, Tracing::class . '@handleStartExecution');
-        $dispatcher->listen(BuildExtensionsResponse::class, Tracing::class . '@handleBuildExtensionsResponse');
```
I had problems when running with laravel octane using the shortcuts

cappuc commented:
I've extracted the `ftv1` string to a constant and replaced it inside the driver and in the headers value of the tests.

I'm not sure it is a good idea to replace it in the test assertions too because it should be replaced event in calls like `$response->json('extensions.ftv1')` if we want to always use the constant and I think it is worst for readability.

I also renamed the value of the config to `federated-tracing` so it is not confused with the header/extension value.

spawnia commented on composer.json:
```diff
@@ -77,7 +78,9 @@
         "mll-lab/laravel-graphiql": "A graphical interactive in-browser GraphQL IDE - integrated with Laravel",
-        "pusher/pusher-php-server": "Required when using the Pusher Subscriptions driver"
+        "pusher/pusher-php-server": "Required when using the Pusher Subscriptions driver",
+        "google/protobuf": "Required when using ftv1 tracing driver",
```
```suggestion
        "google/protobuf": "Required when using the tracing driver federated-tracing",
```

spawnia commented on src/Tracing/TracingServiceProvider.php:
```diff
@@ -25,7 +36,8 @@ public function boot(Dispatcher $dispatcher): void
         ));
-        $dispatcher->listen(StartExecution::class, Tracing::class . '@handleStartExecution');
-        $dispatcher->listen(BuildExtensionsResponse::class, Tracing::class . '@handleBuildExtensionsResponse');
+        $dispatcher->listen(StartRequest::class, static fn (StartRequest $event) => Container::getInstance()->make(Tracing::class)->handleStartRequest($event));
```
```suggestion
        $dispatcher->listen(StartRequest::class, static fn (StartRequest $event) => Container::getInstance()
            ->make(Tracing::class)
            ->handleStartRequest($event));
```

spawnia commented on src/lighthouse.php:
```diff
@@ -453,4 +453,24 @@
+         *
+         * Will be removed in Lighthouse v7 and implicitly default to 'federated-tracing'.
+         */
+        'driver' => env('LIGHTHOUSE_TRACING', \Nuwave\Lighthouse\Tracing\ApolloTracing\ApolloTracing::NAME),
```
```suggestion
        'driver' => env('LIGHTHOUSE_TRACING', Nuwave\Lighthouse\Tracing\ApolloTracing\ApolloTracing::NAME),
```

spawnia replied:
I am not sure an `env` variable is needed, I don't see this changing per environment? We could just use the class names instead then, that would also allow the addition of custom classes because we would no longer need to hardcode driver name to classes.

cappuc replied:
Changing the tracing driver per environment I don't think is too useful but it can be useful for example to enable/disable tracing if we add a `null` driver that does nothing. This can be useful for testing/local environment where tracing is only an overhead. If we remove use the class name directly, this can be implemented with a simple boolean flag.

Another thing to consider is keep the "driver" architecture to allow other drivers to be supported in the future (for example `opentelemetry`)

cappuc replied:
The new federated tracing is enabled only when there is the `apollo-federation-include-trace` header, the old apollo tracing is always enabled.

Anyway conditionally registering the service provider is a good solution for enabling/disabling the tracing.

spawnia commented on tests/Integration/Tracing/FederatedTracingExtensionTest.php:
```diff
@@ -0,0 +1,164 @@
+    {
+        parent::setUp();
+
+        config()->set('lighthouse.tracing.driver', FederatedTracing::NAME);
```
```suggestion
        $config = $this->app->make(ConfigRepository::class);
        $config->set('lighthouse.tracing.driver', FederatedTracing::NAME);
```

spawnia commented on buf.gen.yaml:
```diff
@@ -0,0 +1,7 @@
+# buf.gen.yaml
```
```suggestion
```

cappuc commented:
The workflow should work but I couldn't get it to work on pull request (the main problem is the commit part).
I enabled it only on push like the `format` workflow

cappuc commented on buf.gen.yaml:
```diff
@@ -0,0 +1,6 @@
+managed:
+  enabled: true
+plugins:
+  - plugin: buf.build/protocolbuffers/php
```
@spawnia Do you want to pin a specific protobuf version or always use the latest?

spawnia replied:
Can we pin a major version to avoid incidental breaking changes but still allow for bugfixes?

cappuc replied:
The plugin follows the protobuf version number (and there is a new "major" version every few months but I think they don't follow semver) but usually there are no breaking changes in generated PHP code. 
In the last 4/5 releases that I used they added some phpdoc to improve types

Here are the release pages:
https://buf.build/protocolbuffers/php
https://github.com/protocolbuffers/protobuf/releases

spawnia commented on CONTRIBUTING.md:
```diff
@@ -113,6 +113,14 @@ $user = User::create([
+When updating the proto files, PHP classes needs to be regenerated.
+The generation is done with [buf](https://buf.build/docs/generate/overview).
+The `make proto` command generates the new PHP classes and replace the old ones.
+
```
```suggestion
```

spawnia commented on CONTRIBUTING.md:
```diff
@@ -113,6 +113,14 @@ $user = User::create([
 
+## Working with proto files
+
+Lighthouse uses [protobuf](https://developers.google.com/protocol-buffers) files for federated tracing schema `src/Tracing/FederatedTracing/Schema`.
```
```suggestion
Lighthouse uses [protobuf](https://developers.google.com/protocol-buffers) files for [federated tracing](src/Tracing/FederatedTracing/Schema).
```

spawnia commented on CONTRIBUTING.md:
```diff
@@ -113,6 +113,14 @@ $user = User::create([
+## Working with proto files
+
+Lighthouse uses [protobuf](https://developers.google.com/protocol-buffers) files for federated tracing schema `src/Tracing/FederatedTracing/Schema`.
+When updating the proto files, PHP classes needs to be regenerated.
```
```suggestion
When updating the proto files, the PHP classes need to be regenerated.
```

spawnia commented on src/Tracing/TracingServiceProvider.php:
```diff
@@ -3,19 +3,30 @@
     public function register(): void
     {
-        $this->app->singleton(Tracing::class);
+        $this->app->scoped(Tracing::class, static function (Application $app): Tracing {
+            if (config('lighthouse.tracing.driver') === FederatedTracing::class) {
+                return $app->make(FederatedTracing::class);
+            }
+
+            return $app->make(ApolloTracing::class);
+        });
```
```suggestion
        $this->app->scoped(Tracing::class, static fn (Application $app): Tracing => $app->make(config('lighthouse.tracing.driver')));
```

spawnia commented on src/lighthouse.php:
```diff
@@ -453,4 +453,24 @@
+        /*
+         * Driver used for tracing.
+         *
+         * Available options: '\Nuwave\Lighthouse\Tracing\ApolloTracing\ApolloTracing::class', '\Nuwave\Lighthouse\Tracing\FederatedTracing\FederatedTracing::class'
```
```suggestion
         * Accepts the fully qualified class name of a class that implements Nuwave\Lighthouse\Tracing\Tracing.
         * Lighthouse provides:
         * - Nuwave\Lighthouse\Tracing\ApolloTracing\ApolloTracing::class
         * - Nuwave\Lighthouse\Tracing\FederatedTracing\FederatedTracing::class
```

spawnia commented on src/lighthouse.php:
```diff
@@ -453,4 +453,24 @@
+         *
+         * Will be removed in Lighthouse v7 and implicitly default to 'Nuwave\Lighthouse\Tracing\FederatedTracing\FederatedTracing::class'.
+         */
+        'driver' => \Nuwave\Lighthouse\Tracing\ApolloTracing\ApolloTracing::class,
```
```suggestion
        'driver' => Nuwave\Lighthouse\Tracing\ApolloTracing\ApolloTracing::class,
```

spawnia commented on src/lighthouse.php:
```diff
@@ -453,4 +453,24 @@
+         * Driver used for tracing.
+         *
+         * Available options: '\Nuwave\Lighthouse\Tracing\ApolloTracing\ApolloTracing::class', '\Nuwave\Lighthouse\Tracing\FederatedTracing\FederatedTracing::class'
+         *
+         * Will be removed in Lighthouse v7 and implicitly default to 'Nuwave\Lighthouse\Tracing\FederatedTracing\FederatedTracing::class'.
```
I guess we can still allow overwriting the binding of the scoped singleton for customization then?

cappuc replied:
Yes we can simply change the default in v7 but keeping the driver structure so it allows customization

cappuc commented:
I've added a workflow to automatically update the `reports.proto` file from apollo. I set it to run weekly

spawnia commented on buf.gen.yaml:
```diff
@@ -0,0 +1,6 @@
+managed:
+  enabled: true
+plugins:
+  - plugin: buf.build/protocolbuffers/php:v25.1
```
```suggestion
  # For new releases, see https://github.com/protocolbuffers/protobuf/releases
  - plugin: buf.build/protocolbuffers/php:v25.1
```

cappuc commented on .github/workflows/update-reports-proto.yml:
There is this option to push to protected branches:
https://github.com/stefanzweifel/git-auto-commit-action#push-to-protected-branches

spawnia commented:
Apart from adding the changelog entry and updating the docs to say that federated tracing is supported, I am fine to go ahead and merge this. Excellent work!

Do you plan on updating the Apollo compatibility repository as well? I think we should just need to change `tracing.driver` in the configuration to make the example project there work?

cappuc commented:
We should take a look at error reporting (I wrote some notes in the first comment).

Apart from that and changelog/docs update it, everything should be fine.

Yes I think the only required change for apollo compatibility is the change of `tracing.driver`. I will make a PR to update apollo compatibility when this PR is merged.

spawnia commented:
I am not sure how error reporting in tracing is/should be different from error reporting for clients. Does https://lighthouse-php.com/master/digging-deeper/error-handling.html apply? It would be nice to keep things simple and just utilize the same mechanisms. See `\Nuwave\Lighthouse\GraphQL::errorsHandler()`.

cappuc commented:
Yes it apply all handlers.
I just changed the message to be "Internal server error" when error is not client safe and removed the json data, so we provide only message and location

spawnia commented on docs/master/federation/getting-started.md:
```diff
@@ -50,6 +50,13 @@ extend schema @link(
+## Federated tracing
+
+In order to use federated tracing, you need to enabled [tracing](../performance/tracing.md)
+and set the driver to `Nuwave\Lighthouse\Tracing\FederatedTracing\FederatedTracing::class` in your `config/lighthouse.php`:
```
Were you planning on adding example code here, or should we end this with `.`?

spawnia commented on docs/master/performance/tracing.md:
```diff
@@ -12,3 +11,17 @@ Add the service provider to your `config/app.php`
+
+## Drivers
+
+Lighthouse tracing is implemented though drivers, this allows to easily support different tracing formats.
```
```suggestion
Lighthouse tracing is implemented though drivers, this allows supporting different tracing formats.
```

https://justsimply.dev 😉

spawnia commented:
Ready to merge and release?

cappuc commented:
Yes, I think it's ready

Maxwell2022 commented:
We need to update the doc too: https://lighthouse-php.com/6/federation/getting-started.html#federated-tracing

> Experimental: not enabled by default, not guaranteed to be stable.

Is this still experimental?
